utils/upgrade

```diff
@@ -0,0 +1,345 @@
+
+    def do_work(self):
+        if self.state_machine.params["force"]:
+            self.will_prompt = False
+        elif yn_prompt("Proceed with upgrade procedure?") == "n":
+            return Abort("User aborted")
+
+        try:
+            version = opencas.get_cas_version()
+        except FileNotFoundError:
+            return self.NotInstalled(f"Couldn't detect current CAS version")
+        except Exception as e:
+            return Failure(f"Failed to get current version of CAS {e}")
+
+        if version[CAS_CLI_KEY] != version[CAS_CACHE_KEY]:
+            return Failure("Mismatch between CLI and cas_cache version")
+
+        active_devices = opencas.get_devices_state()
+
+        if any([device["status"] != "Running" for _, device in active_devices["caches"].items()]):
+            return Failure("Incomplete configuration. Run casadm -L to review CAS state!")
+
+        if len(active_devices["core_pool"]) != 0:
+            return Failure("Incomplete configuration. Run casadm -L to review CAS state!")
+
+        return Success()
+
+
+class BuildCas(UpgradeState):
+    log = "Compiling Open CAS"
+
+    def do_work(self):
+        with open(COMPILATION_LOG, "w") as build_log:
+            logging.info("Running ./configure for CAS")
+            p = subprocess.run(["./configure"], cwd="..", stdout=build_log, stderr=build_log)
+            if p.returncode:
+                return Failure(f"Configuration of Open CAS failed. Build log: {COMPILATION_LOG}")
+
+            logging.info("Compiling CAS")
+            p = subprocess.run(["make", "-j"], cwd="..", stdout=build_log, stderr=build_log)
+            if p.returncode:
+                return Failure(f"Compilation of Open CAS failed. Build log: {COMPILATION_LOG}")
+
+        return Success()
+
+
+class InstallCas(UpgradeState):
+    log = "Installing new Open CAS files"
+
+    def do_work(self):
+        with open(COMPILATION_LOG, "a") as install_log:
+            p = subprocess.run(
+                ["make", "install"], cwd="..", stdout=install_log, stderr=install_log
+            )
+
+            if p.returncode:
+                return Failure(
+                    f"Installation of Open CAS failed. Installation log: {COMPILATION_LOG}"
+                )
+
+        return Success()
+
+
+class InsertModule(UpgradeState):
+    module_path = f"{os.path.dirname(__file__)}/../modules/cas_cache/cas_cache.ko"
+    options = {"installed": False}
+
+    def do_work(self):
+        try:
+            insert_module(self.module_path, **self.options)
+        except Exception as e:
+            return Failure(f"Couldn't load module {self.module_path}. Reason: {e}")
+
+        return Success()
+
+
+class InsertNewModule(InsertModule):
+    log = "Try to insert new caching module"
+
+
+class DryRun(InsertModule):
+    log = "Perform dry run to check upgrade data integrity"
+    options = {"installed": False, "dry_run": 1}
+
+    def do_work(self):
+        result = super().do_work()
```

Comment:
Wouldn't this be the place to workaround checking available/free memory?

Reply:
Included workaround in a separate state